src/libzbxpgsql.c

```diff
@@ -567,3 +620,286 @@ char *strcat2(char *dest, const char *src)
+                if (fileCount >= MAX_NUMBER_CONFIG_FILES) {
+                    zabbix_log(LOG_LEVEL_ERR, "%s ERROR: Found more than %i config files: %i", PACKAGE, MAX_NUMBER_CONFIG_FILES, fileCount);
+                    return EXIT_FAILURE;
+                }
+                strsize = strlen(filenames.gl_pathv[fileCount]) + 1;
+                configPath[fileCount] = zbx_malloc(configPath[fileCount],strsize * sizeof(char));
+                if (configPath[fileCount] == NULL) {
+                    zabbix_log(LOG_LEVEL_CRIT, "%s: ERROR: zbx_malloc failed", PACKAGE);
+                    return EXIT_FAILURE;
+                }
+                zbx_strlcpy(configPath[fileCount],filenames.gl_pathv[fileCount],strlen(filenames.gl_pathv[fileCount])+1);
+            }
+            globfree(&filenames);
+            break;
+        case GLOB_NOMATCH :
+            zabbix_log(LOG_LEVEL_INFORMATION, "%s: No config files found to process... skipping", PACKAGE);
+            fileCount = 0;
+            break;
+        default : // untrapped glob error
+            fileCount = -1;
+    }
+    zabbix_log(LOG_LEVEL_DEBUG, "%s: End of %s (fileCount=%i)", PACKAGE, __function_name, fileCount);
+    return fileCount;
+}
+
+/*
+ * Function storeSQLstmt
+ *
+ * A simple key/value store using arrays of pointers.
+ *
+ * Stores a string key and a string value (SQL statement)
+ * into a sorted array, maintaining sort order.
+ *
+ * Since it's just moving pointers it should perform well
+ * even with large numbers of keys.
+ *
+ * Returns: 
+ *   EXIT_SUCCESS = successfully inserted
+ *   EXIT_FAILURE = insert failed
+ *   -1           = duplicate key discarded
+ */
+int  storeSQLstmt(const char *key, const char *stmt) {
+    const char  *__function_name = "storeSQLstmt";
+    int  i;
+
+    zabbix_log(LOG_LEVEL_DEBUG, "%s: In %s(%s,<stmt>)", PACKAGE, __function_name, key);
+    // make sure we have space
+    if (SQLcount >= MAX_NUMBER_SQL_STATEMENT_IN_RAM) {
+        zabbix_log(LOG_LEVEL_ERR, "%s: ERROR: Keystore full: %i statements stored already", PACKAGE, SQLcount);
+        return EXIT_FAILURE;
+    }
+    // exclude dupes
+    if (SQLkeysearch((char *)key) != -1) {
+        zabbix_log(LOG_LEVEL_ERR, "%s: ERROR: Duplicate key: \"%s\"", PACKAGE, key);
+        return EXIT_FAILURE;
+    }
+    // start at the end of the index and push out
+    // entries to the next spot until you find the
+    // right spot to insert the new key/value pair
+    i = SQLcount - 1;
+    zabbix_log(LOG_LEVEL_TRACE, "%s: Starting to look for insert location (i:%i)", PACKAGE, i);
+    while (i >= 0 && strcmp(key,SQLkey[i]) < 0) {
+        zabbix_log(LOG_LEVEL_TRACE, "%s: moving data from slot %i to slot %i", PACKAGE, i, i+1);
+        SQLkey[i+1]  = SQLkey[i];
+        SQLstmt[i+1] = SQLstmt[i];
+        i--;
+    }
+    // allocate memory for the new key
+    SQLkey[i+1] = zbx_malloc(SQLkey[i+1],sizeof(char) * (strlen(key)+1));
+    if (SQLkey[SQLcount] == NULL) {
```

Comment:
Should be SQLkey[i+1].

I'll fix it and redo the pull request.

Thanks!

On Sat, Aug 27, 2016 at 10:37 PM, Ryan Armstrong <notifications@github.com>
wrote:

> In src/libzbxpgsql.c
> <https://github.com/cavaliercoder/libzbxpgsql/pull/55#discussion_r76525591>
> :
>
> > +        return EXIT_FAILURE;
> > +    }
> > +    // start at the end of the index and push out
> > +    // entries to the next spot until you find the
> > +    // right spot to insert the new key/value pair
> > +    i = SQLcount - 1;
> > +    zabbix_log(LOG_LEVEL_TRACE, "%s: Starting to look for insert location (i:%i)", PACKAGE, i);
> > +    while (i >= 0 && strcmp(key,SQLkey[i]) < 0) {
> > +        zabbix_log(LOG_LEVEL_TRACE, "%s: moving data from slot %i to slot %i", PACKAGE, i, i+1);
> > +        SQLkey[i+1]  = SQLkey[i];
> > +        SQLstmt[i+1] = SQLstmt[i];
> > +        i--;
> > +    }
> > +    // allocate memory for the new key
> > +    SQLkey[i+1] = zbx_malloc(SQLkey[i+1],sizeof(char) * (strlen(key)+1));
> > +    if (SQLkey[SQLcount] == NULL) {
>
> Should this check SQLkey[SQLcount] or SQLkey[i+1]?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cavaliercoder/libzbxpgsql/pull/55/files/79848f2d016dd66c192d4d15c8fc91aae7a1d909#r76525591>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AOBRoN1uAnJb-bXqwNWibIEzZsKSDgiKks5qkQJzgaJpZM4Ju1Cu>
> .
>